Add `role="text"` to Author `span` elements

tylersticka opened the pull request:
## Overview

> Add `role="text"` to `span` elements inside of the Author component, increasing the likelihood that its contents will be read aloud seamlessly

## Testing

[On deploy preview](https://deploy-preview-1982--cloudfour-patterns.netlify.app/?path=/docs/components-author--basic-usage), test VO experience

---

- Fixes #1981

tylersticka commented:
@gerardo-rodriguez @calebeby I know tests are failing, but I've requested a review because the accessibility tree comparison has me a little... I dunno... worried this is the wrong move? The tree _before_ this change makes more sense to me than the tree after, which seems to have nested `text` elements?

So I guess I'd just love another pair of 👀 making sure that I'm not missing something critical here. If I just need to update the test, that's fine, I'm just skittish now.

calebeby commented:
Hmm. I agree that the accessibility tree looks a little funny in the test. I tested this PR in VO and it definitely feels like an improvement.

The pleasantest accessibility tree feature uses `text` to denote text that is not associated with an element's `name`. It'll also use `text` to denote the `text` role. I _think_ there is code to avoid printing text if it is already displayed as an element's name, but I'm not sure. I'll look into this more, but for now I'd just let Jest update the test even though it looks funny.

calebeby commented:
@tylersticka can we add a space between "by" and the author's name? In VO it was reading as ByAuthorfirstname Authorlastname for ones without links. (This is an issue on `main` too but I figure it is worth fixing here while we're at it).

tylersticka commented:
> can we add a space between "by" and the author's name?

Will spaces actually solve the problem? There are already newlines between the elements, which traditionally get collapsed into spaces in HTML.

calebeby commented:
@tylersticka I think there are not newlines in the HTML because the twig tags have the dashes which eat the whitespace

calebeby commented:
> @tylersticka I think there are not newlines in the HTML because the twig tags have the dashes which eat the whitespace

Nevermind, I see that there are some spaces not being eaten by twig. Hmm, I wonder why it is collapsing the words 🤔

tylersticka commented:
@calebeby I just tweaked the dashes and locally in Safari it's being read as I'd expect it to. Will push that along with a text update soon.

calebeby commented on src/components/author/author.twig:
```diff
@@ -63,14 +63,16 @@
             <a class="c-author__author" href="{{ author.link }}">{{ author.name }}</a>
           {%- else -%}
-            <span class="c-author__author">{{ author.name }}</span>
+            <span class="c-author__author" role="text">{{ author.name }}</span>
```
Hmm, if we add a space here it reads "by" and the first name as two separate words:

```suggestion
            <span class="c-author__author" role="text"> {{ author.name }}</span>
```

But, it also messes with it visually, so I'm not sure this is what we want

gerardo-rodriguez commented on src/components/author/author.twig:
```diff
@@ -12,7 +12,7 @@
       <p>
         {# This creates a more accessible, and less confusing, user experience #}
-        <span class="u-hidden-visually">
+        <span class="u-hidden-visually" role="text">
```
We shouldn't need the addition of `role="text"` here since it's one `span` that contains the full text. 😉 
```suggestion
        <span class="u-hidden-visually">
```

tylersticka replied:
@gerardo-rodriguez Can you elaborate on what you mean by "contains the full text?"

gerardo-rodriguez replied:
@tylersticka Sure, sorry for not being clear. 😬 
```twig
{# 
  All of the text that is expected to be read together is within a single span element, so we 
  don't need to add the role="text" workaround to have it read as one whole unit of text 
#}
<span class="u-hidden-visually">
  {{ date_prefix|default('Published on') }} {{ date|date('F jS, Y') }}
</span>
```

gerardo-rodriguez replied:
Am I being clear? I can try again if I still didn't explain myself well. 😅

tylersticka replied:
Okay, I think I get it. If the element is an "only child," there's no _gap_ between elements the user has to manually traverse. If there were an additional `span` after this one, then there would be.

Thank you for the explanation.

calebeby commented:
It looks like role="text" is nonstandard. It provides a usability improvement for VO+safari users, so I'm not opposed to using it if we don't have a better option, but just thought I would mention that it was nonstandard in case you didn't know.

Draft spec: https://w3c.github.io/aria/
Published spec: https://www.w3.org/TR/wai-aria/
More details on role="text": https://www.tpgi.com/using-the-text-role/

tylersticka commented:
Hmm, this paragraph is really standing out to me:

> To a developer or user who is used to the way JAWS or NVDA announces content, the behavior of VO might seem fragmented and annoying, but to someone who regularly uses VO, this pattern of speech output is usual, and may be the expected or even preferred pattern.

@gerardo-rodriguez I fear we may be overstepping to bring VO usability in line with other tools. I'm going to shift this back to "draft" for now and consider this a bit further.

gerardo-rodriguez commented:
> It looks like role="text" is nonstandard. It provides a usability improvement for VO+safari users, so I'm not opposed to using it if we don't have a better option, but just thought I would mention that it was nonstandard in case you didn't know.

I was just reading this thread: https://github.com/dequelabs/axe-core/issues/1597

gerardo-rodriguez commented:
> Hmm, this paragraph is really standing out to me:
> 
> > To a developer or user who is used to the way JAWS or NVDA announces content, the behavior of VO might seem fragmented and annoying, but to someone who regularly uses VO, this pattern of speech output is usual, and may be the expected or even preferred pattern.
> 
> @gerardo-rodriguez I fear we may be overstepping to bring VO usability in line with other tools. I'm going to shift this back to "draft" for now and consider this a bit further.

@tylersticka Totally fair! We haven't had the opportunity to have actual user testing so we are making _a lot_ of assumptions to try and provide what we think is the best UX based on our best intentions.

gerardo-rodriguez commented:
> I fear we may be overstepping to bring VO usability in line with other tools. 

Though, it _could_ be argued that if it leads to a better end-user experience, it's worth it. (Priority of constituencies?) Why not push the status quo?

But, I guess again, I'm making an assumption that a VO (macOS or iOS) end-user prefers not needing to have VO read/navigate the broken-up text. 😕

tylersticka commented:
@gerardo-rodriguez @calebeby After re-reading both threads and playing more with the VO experience, I think this is a valid use of `role="text"`.

It's not a "home run" sort of decision. Clearly, Safari has decided that these `span` elements are worthy of separate consideration. If a user has chosen VoiceOver as their assistive device, then surely they must be used to that sort of distinction.

I just think in this case it's very clearly not useful. The reason the `span` elements exist is entirely visual: Hiding text that's unnecessary in a larger visual context and preventing line breaks. They are not _emphasizing_ content, they are not grouping it in a meaningful way… it's purely presentational.

That said, from a development standpoint this seems pretty ridiculous. I think there's a limit to how many attributes we should add to every single adjacent `span` that we could consider reasonable. But in this case, we're talking about two `span` elements in a template that seem to make the experience better. I think we should continue to judge this sort of thing on a case by case basis.

Per @gerardo-rodriguez's earlier comment, I've removed the unnecessary `role` from the `time` fallback text (and updated the related test to match). I've also added a large comment with this context preceding the part of the template with `role="text"`.

If you disagree with this conclusion, let me know and we can chat more about it with voices. Thanks for bearing with me on this!

gerardo-rodriguez commented:
To add to the conversation. 🙃 

<img width="758" alt="Screen Shot 2022-07-26 at 4 05 45 PM" src="https://user-images.githubusercontent.com/459757/181127647-1e4ccae0-c982-494a-933c-61b2a6837b4f.png">

Sorry to drag you along this journey. 
- I was navigating the site with VO
- I was annoyed when text was split up when it was read to me
- I wondered if I could do anything about that
- I found this: https://axesslab.com/text-splitting/
- It worked
- I shared the possibility of potentially enhancing the UX
- And here we are 

I appreciate ya'll digging more.

I still don't know what the "right" answer is. I guess, if I step back, there is no right answer because we are talking about UX. It's subjective and as @tylersticka mentioned, should be judged on a case by case basis.

tylersticka commented:
If Adrian says we shouldn't do this, then I'm going to go with what he says. 😅
